File: BOJ/2109/2169.py
# https://www.acmicpc.net/problem/2169
# DP
from sys import stdin

INF = 10 ** 10
n, m = map(int, stdin.readline().split())
seq = [list(map(int, stdin.readline().split())) for _ in range(n)]
dp_left = [[-INF for _ in range(m)] for _ in range(n)]
dp_right = [[-INF for _ in range(m)] for _ in range(n)]
dp = [[-INF for _ in range(m)] for _ in range(n)]

for col in range(1, m):
    seq[0][col] += seq[0][col - 1]
dp_left[0] = seq[0]
dp_right[0] = seq[0]
dp[0] = seq[0]

for row in range(1, n):
    for col in range(m):
        dp_left[row][col] = max(dp_left[row][col - 1], dp[row - 1][col]) + seq[row][col]
    for col in range(-1, -m - 1, -1):
        dp_right[row][col] = max(dp_right[row][col + 1], dp[row - 1][col]) + seq[row][col]
    for col in range(m):
        dp[row][col] = max(dp_left[row][col], dp_right[row][col])

# for row in range(n):
#     print(dp[row])

print(dp[-1][-1])

File: BOJ/2109/2169.cpp
// O(n^3) but it passes!

#include <iostream>

int main()
{
    int n, m;
    int INF = 1000000000;
    scanf("%d%d", &n, &m);
    int seq[n][m];
    int dp[n][m];

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            int foo;
            scanf("%d", &foo);
            seq[i][j] = foo;
            dp[i][j] = -INF;
        }
    }

    int foo = 0;
    for (int i = 0; i < m; i++)
    {
        foo += seq[0][i];
        dp[0][i] = foo;
    }

    for (int row = 0; row < n - 1; row++)
    {
        for (int col = 0; col < m; col++)
        {
            int partial = dp[row][col];
            for (int x = col; x > -1; x--)
            {
                partial += seq[row + 1][x];
                if (partial > dp[row + 1][x])
                    dp[row + 1][x] = partial;
            }
            partial = dp[row][col];
            for (int x = col; x < m; x++)
            {
                partial += seq[row + 1][x];
                if (partial > dp[row + 1][x])
                    dp[row + 1][x] = partial;
            }
        }
    }

    printf("%d", dp[n - 1][m - 1]);

    // for (int i = 0; i < n; i++)
    // {
    //     for (int j = 0; j < m; j++)
    //     {
    //         printf("%d/", seq[i][j]);
    //     }
    //     printf("\n");
    // }
    // printf("================================\n");
    // for (int i = 0; i < n; i++)
    // {
    //     for (int j = 0; j < m; j++)
    //     {
    //         printf("%d/", dp[i][j]);
    //     }
    //     printf("\n");
    // }
}